npm_and_yarn/helpers/lib/npm/vulnerability-auditor.js

```diff
@@ -148,21 +149,44 @@ function convertAdvisoriesToRegistryBulkFormat(advisories) {
  * applies to the first item in the chain (if that item is fixable, then
  * every item after it must be fixable, too).
  */
-function buildDependencyChains(auditReport, name, chain = { items: [] }) {
-  const vuln = auditReport.get(name)
-  const version = [...vuln.nodes][0].version
-  const item = { name, version }
-
-  if (!vuln.effects.size) {
-    // If the current vuln has no effects, we've reached the end of this chain.
-    return [{ fixAvailable: vuln.fixAvailable, items: [item, ...chain.items] }]
+function buildDependencyChains(auditReport, name) {
+  const helper = (name, chain) => {
+    // The vuln for this dependency.
+    const vuln = auditReport.get(name)
+
+    // The current version of this dependency.
+    const version = [...vuln.nodes][0].version
+
+    // The item that will represent this dependency in this chain.
+    const item = { name, version }
+
+    // Set containing the names of all dependencies in this chain up to and including this one;
+    // used for preventing cycles.
+    const seen = new Set([name, ...chain.items.map(i => i.name)])
+
+    // Array of effects, excluding cycles.
+    const effects = [...vuln.effects].filter(effect => !seen.has(effect.name))
```

Comment:
Could an effect have the same name but a different vulnerability? We might need to compare the vulns as well if that's the case.

Reply:
Since we're starting from a single dependency, it won't matter whether it has 1 vuln associated with it or 10—its `effects` are simply the names of other dependencies that are vulnerable as a result of depending on it, and so on up the chain.